pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -323,29 +334,41 @@ func makeUID() string {
 	return fmt.Sprintf("%08x", rand.Uint32())
 }
 
-// getTerminationMessage gets termination message of the container.
-func getTerminationMessage(status *runtimeapi.ContainerStatus, kubeStatus *kubecontainer.ContainerStatus, terminationMessagePath string) string {
-	message := ""
-
-	if !kubeStatus.FinishedAt.IsZero() || kubeStatus.ExitCode != 0 {
-		if terminationMessagePath == "" {
-			return ""
-		}
+const maxMessageLength = 1024 * 4
 
+// getTerminationMessage looks on the filesystem for the provided termination message path, returning a limited
+// amount of those bytes, or returns true if the logs should be checked.
+func getTerminationMessage(status *runtimeapi.ContainerStatus, terminationMessagePath string, fallbackToLogs bool) (string, bool) {
+	if len(terminationMessagePath) != 0 {
 		for _, mount := range status.Mounts {
-			if mount.GetContainerPath() == terminationMessagePath {
-				path := mount.GetHostPath()
-				if data, err := ioutil.ReadFile(path); err != nil {
-					message = fmt.Sprintf("Error on reading termination-log %s: %v", path, err)
-				} else {
-					message = string(data)
-				}
-				break
+			if mount.GetContainerPath() != terminationMessagePath {
+				continue
+			}
+			path := mount.GetHostPath()
+			data, _, err := tail.ReadAtMost(path, maxMessageLength)
+			if err != nil {
+				return fmt.Sprintf("Error on reading termination log %s: %v", path, err), false
+			}
+			if !fallbackToLogs || len(data) != 0 {
+				return string(data), false
 			}
 		}
 	}
+	return "", fallbackToLogs
+}
 
-	return message
+// readLastStringFromContainerLogs attempts to read up to maxMessageLength from the end of the CRI log represented
+// by path. It reads up to maxMessageLength lines (on the assumption every line is at least one byte).
+func readLastStringFromContainerLogs(path string) string {
+	logOptions := &v1.PodLogOptions{}
+	// TODO: more efficient byte search option
+	value := int64(maxMessageLength)
+	logOptions.TailLines = &value
```

Comment:
LimitBytes is from the beginning of the logs, while TailLines is from the end.  If  we didn't want to avoid introducing a new API to LogOptions, I'd suggest TailBytes because it's fairly useful.  Unfortunately for here we have to get lines > bytes, then use a circular buffer to purge the extra bytes.